design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,474 @@
+
+This is problematic for us at this stage because now the Secrets Provider is sitting separately and will need to perform a batch retrieval of all the secrets in the namespace. In other words, if the Secrets Provider fails to fetch one Conjur secret, the whole request will fail and applications will not get their secrets. 
+
+To avoid this, we will parse over each K8s Secret and perform a batch retrieval request. That way, if there is a failure in retrieving a Conjur secret (a failure in permissioning for example), it will be contained to that secret and not impact all the secrets in the namespace.
+
+```pseudocode
+for each k8s-secret
+  variable-ids = k8s-secret.parseConjurMap() # list of paths of key in ConjurMap
+  GET /secrets{?variable-ids}
+  # Log failure or success
+```
+
+Variable IDs will resemble the following:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-credentials
+type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets-accessors/db_username # the value is one element in the variable-ids list
+    password: secrets-accessors/db_password # the value is one element in the variable-ids list
+```
+
+#### Fetching Relevant K8s Secrets
+
+With the decoupling of Secrets Provider and application pod, we can know what types of K8s Secrets the pod requires by either of the following ways: 
+
+|      | Fetch/Update K8s Secret                                      | Pros                                                         | Cons                                        | Effort estimation |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------- | ----------------- |
+| 1    | Read all K8s Secrets that are marked with labels specified in `K8S_SECRETS_LABEL` env var | - Scalable <br /><br />- Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound | 3 days            |
+| 2    | Get K8s secrets with the names specified in `K8S_SECRETS`    | - No code changes<br /><br />- Backwards compatible          | - Repetitive work<br /><br />- Not scalable | Free              |
+
+*Decision*: Solution #1, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs.
+
+Listing each K8s Secret in the ENV of the Secrets Provider manifest is not a scalable solution, especially for customers with hundreds of K8s Secrets. Customers might already use labels for grouping related K8s Secrets so we will benefit from this enhancement by integrating more with a customer environment and utilizing that.
```

Comment:
but is it scalable to run around all the secrets to add a label?

Reply:
The idea behind this design was to better integrate with customer environments because they may already be using  labels so they don't have to add entries to every secret

Reply:
did we verify this with customers? do they have such labels that they can use? let's not develop a feature that will not be used.

Reply:
Completely agree @InbalZilberman gave a 👍